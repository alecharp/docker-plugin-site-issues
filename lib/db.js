const axios = require('axios');
const config = require('./config.js');

const quoteStr = (str) => `"${str.replace('"', '\\"')}"`;

async function getPluginInfo(pluginName) {
  const response = await axios.get('https://updates.jenkins.io/update-center.actual.json').then((response) => response.data);
  if (response.plugins.hasOwnProperty(pluginName)) {
    return response.plugins[pluginName];
  }
}

async function getIssuesForPlugin(pluginName) {
  const response = await axios.get('https://reports.jenkins.io/issues.index.json').then(response => response.data);
  if (response.hasOwnProperty(pluginName)) {
    return response[pluginName];
  }
}

async function getJiraIssues(componentId, statuses = ['Open', 'In Progress', 'Reopened']) {
  const maxResults = 100;
  const startAt = 0;

  const axiosOptions = {
    method: 'get',
    url: `${config.jira.url.replace(/\/+$/, '')}/rest/api/2/search`,
    auth: {
      username: config.jira.username,
      password: config.jira.password
    },
    params: {
      startAt,
      maxResults,
      jql: `project=JENKINS AND status in (${statuses.map(quoteStr).join(',')}) AND component=${componentId}`
    }
  };
  const response = await axios(axiosOptions).then(response => response.data);
  return response;
}

module.exports = {
  getPluginInfo,
  getIssuesForPlugin,
  getJiraIssues
};
