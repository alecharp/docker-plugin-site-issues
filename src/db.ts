import axios from 'axios';
import { Octokit } from '@octokit/rest';
import { createAppAuth } from '@octokit/auth-app';
import { request as githubRequest } from '@octokit/request';
import JiraApi from 'jira-client';
import { MemoizeExpiring } from 'typescript-memoize';
import NodeCache from 'node-cache';

import config from './config.js';

const DEFAULT_TIMEOUT = 60 * 60 * 1000;

function quoteStr(str: string) {
  return `"${str.replace('"', '\\"')}"`;
}

const cache = new NodeCache({
  stdTTL: DEFAULT_TIMEOUT,
});

export type Issue = {
  readonly key: string;

  readonly issueTypes: string[];

  readonly priority: string;

  readonly status: string;

  readonly resolution: string;

  readonly summary: string;

  readonly assignee: string;

  readonly reporter: string;

  readonly created: string;

  readonly updated: string;

  readonly url: string;
};

type IssueIndex = {
  readonly type: string;
  readonly reference: string;
  readonly viewUrl?: string;
  readonly reportUrl?: string;
};

export interface IDB {
  getIssuesForPlugin(pluginName: string) : Promise<Array<IssueIndex> | null>;
  getJiraIssues(component: number | string, startAt: number, statuses: Array<string>): Promise<Issue[]>;
  getGithubIssues(reference: string): Promise<Issue[]>;
}

export class DB implements IDB {
  @MemoizeExpiring(DEFAULT_TIMEOUT)
  async getIssuesForPlugin(pluginName: string) : Promise<Array<IssueIndex> | null> {
    let pluginIssuesIndex : Record<string, Array<IssueIndex>> | undefined = cache.get('issues.index.json');
    if (!pluginIssuesIndex) {
      pluginIssuesIndex = await axios.get('https://reports.jenkins.io/issues.index.json')
        .then((response) => response.data || {});
      cache.set('issues.index.json', pluginIssuesIndex);
    }
    if (pluginIssuesIndex && Object.prototype.hasOwnProperty.call(pluginIssuesIndex, pluginName)) {
      return pluginIssuesIndex[pluginName];
    }
    return null;
  }

  @MemoizeExpiring(DEFAULT_TIMEOUT, (component, startAt, statues) => JSON.stringify([component, startAt, statues]))
  async getJiraIssues(component: number | string, startAt = 0, statuses = ['Open', 'In Progress', 'Reopened']): Promise<Issue[]> {
    const maxResults = 1000;

    const parsedUrl = new URL(config.jira.url);
    const jira = new JiraApi({
      protocol: parsedUrl.protocol,
      host: parsedUrl.host,
      port: parsedUrl.port,
      base: parsedUrl.pathname.replace(/^\/+/g, ''),
    });

    const response = await jira.searchJira(
      `project=JENKINS AND status in (${statuses.map(quoteStr).join(',')}) AND component=${component}`, {
        startAt,
        maxResults,
        fields: ['key', 'components', 'summary', 'assignee', 'reporter', 'priority', 'status', 'resolution', 'created', 'updated'],
      },
    );

    const issues = response.issues.map((issue: any) => ({
      key: issue.key,
      issueTypes: [issue?.fields?.issuetype?.name],
      priority: issue?.fields?.priority?.name,
      status: issue?.fields?.status?.status,
      resolution: issue?.fields?.resolution?.name,
      summary: issue?.fields?.summary,
      assignee: issue?.fields?.assignee?.displayName,
      reporter: issue?.fields?.reporter?.displayName,
      created: issue?.fields?.created,
      updated: issue?.fields?.updated,
      url: `${config.jira.url.replace(/\/+$/, '')}/browse/${issue.key}`,
    }));
    if (response.startAt + response.issues.length < response.total) {
      issues.push(...await this.getJiraIssues(component, startAt + maxResults, statuses));
    }
    return issues;
  }

  @MemoizeExpiring(DEFAULT_TIMEOUT)
  private async getRestClient() {
    const { data: installations } = await createAppAuth({
      appId: config.github.appId,
      privateKey: config.github.privateKey,
    }).hook(githubRequest.defaults({
      baseUrl: config.github.server,
    }), 'GET /app/installations');

    return new Octokit({
      baseUrl: config.github.server,
      request: {
        hook: createAppAuth({
          appId: config.github.appId,
          privateKey: config.github.privateKey,
          installationId: installations[0].id,
        }).hook,
      },
    });
  }

  @MemoizeExpiring(DEFAULT_TIMEOUT)
  async getGithubIssues(reference: string): Promise<Issue[]> {
    const [owner, repo] = reference.split('/');

    const octokit = await this.getRestClient();
    const iterator = octokit.paginate.iterator(octokit.rest.issues.listForRepo, {
      per_page: 100,
      owner,
      repo,
      state: 'open',
    });

    const returnedIssues : Array<Issue> = [];
    for await (const { data: issues } of iterator) {
      for (const issue of issues) {
        let labels: string[] = [];
        if (issue.labels && Array.isArray(issue.labels) && issue.labels.length > 0) {
          labels = issue.labels.map((n: any) => {
            if (typeof n === 'string') {
              return n;
            }

            return n.name || '';
          });
        }
        returnedIssues.push({
          key: `#${issue.number}`,
          issueTypes: labels,
          priority: '',
          resolution: issue.state,
          status: issue.state,
          summary: issue.title,
          assignee: issue.assignees?.map((n: any) => n?.login || '').join(', ') || '',
          reporter: issue.user?.login || '',
          created: issue.created_at,
          updated: issue.updated_at,
          url: issue.html_url,
        });
      }
    }
    return returnedIssues;
  }
}
