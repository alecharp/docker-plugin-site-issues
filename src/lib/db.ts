import axios, { Method } from 'axios';
import config from './config';

const quoteStr = (str: string) => `"${str.replace('"', '\\"')}"`;

async function getPluginInfo(pluginName: string) {
  const response = await axios.get('https://updates.jenkins.io/update-center.actual.json').then((response) => response.data);
  if (response.plugins.hasOwnProperty(pluginName)) {
    return response.plugins[pluginName];
  }
}

async function getIssuesForPlugin(pluginName: string) {
  const response = await axios.get('https://reports.jenkins.io/issues.index.json').then((response) => response.data);
  if (response.hasOwnProperty(pluginName)) {
    return response[pluginName];
  }
}

async function getJiraIssues(componentId: number, statuses = ['Open', 'In Progress', 'Reopened']): Promise<any> {
  const maxResults = 100;
  const startAt = 0;

  const axiosOptions = {
    method: 'get' as Method,
    url: `${config.jira.url.replace(/\/+$/, '')}/rest/api/2/search`,
    auth: {
      username: config.jira.username,
      password: config.jira.password,
    },
    params: {
      startAt,
      maxResults,
      jql: `project=JENKINS AND status in (${statuses.map(quoteStr).join(',')}) AND component=${componentId}`,
    },
  };
  // TODO - loop through all the pages
  // if (obj.getInt("startAt") + jsonIssues.length() < obj.getInt("total")) {
  //   jiraIssues.issues.addAll(getIssues(pluginName, startAt + maxResults).issues);
  // }
  const response = await axios(axiosOptions).then((response) => response.data);
  return response.issues.map((issue: { key: string; fields: { issuetype?: { name?: string; }; priority?: { name?: string; }; status?: { status?: string; }; resolution: { name?: string; }; summary?: string; assignee?: { displayName?: string; }; reporter?: { displayName?: string; }; created: string; updated: string; }; }) => ({
    key: issue.key,
    issueType: issue?.fields?.issuetype?.name,
    priority: issue?.fields?.priority?.name,
    status: issue?.fields?.status?.status,
    resolution: issue?.fields?.resolution?.name,
    summary: issue?.fields?.summary,
    assignee: issue?.fields?.assignee?.displayName,
    reporter: issue?.fields?.reporter?.displayName,
    created: issue?.fields?.created,
    updated: issue?.fields?.updated,
    url: `${config.jira.url.replace(/\/+$/, '')}/browse/${issue.key}`,
  }));
}

export default {
  getPluginInfo,
  getIssuesForPlugin,
  getJiraIssues,
};
