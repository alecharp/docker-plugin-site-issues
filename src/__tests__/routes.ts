import test from 'ava';
import httpMocks from 'node-mocks-http';
import {
  indexRoute,
  healthcheckRoute,
  infoRoute,
  issuesRoute,
} from '../routes.js';
// @ts-ignore
import { setupTests } from '../ava-nock/src/index.js';
import { DB } from '../db.js';

setupTests(test);

test('routes > infoRoute > should return 200 with commit and version', async (t) => {
  const req = httpMocks.createRequest();
  req.db = new DB();
  const res = httpMocks.createResponse();

  infoRoute(req, res);

  const data = res._getJSONData();
  t.is(200, res.statusCode);

  t.true(res._isEndCalled());
  t.true(res._isJSON());
  t.true(res._isUTF8());

  t.truthy(data.commit);
  t.truthy(data.version);
});

test('routes > healthcheckRoute > should return 200', async (t) => {
  const req = httpMocks.createRequest();
  req.db = new DB();
  const res = httpMocks.createResponse();

  healthcheckRoute(req, res);

  t.is('OK', res._getData());
  t.is('text/plain', res.getHeader('Content-Type') as string);
});

test('routes > indexRoute > should return 200', async (t) => {
  const req = httpMocks.createRequest();
  req.db = new DB();
  const res = httpMocks.createResponse();

  indexRoute(req, res);

  t.is('OK', res._getData());
  t.is('text/plain', res.getHeader('Content-Type') as string);
});

test('routes > issuesRoute > should return 200', async (t) => {
  const req = httpMocks.createRequest({ params: { plugin: 'configuration-as-code' } });
  req.db = new DB();
  const res = httpMocks.createResponse();

  await issuesRoute(req, res);

  t.is(200, res.statusCode);
  t.true(res._isJSON());
  t.true(res._isUTF8());
  t.snapshot(res._getJSONData());
});

test('routes > issuesRoute > should return 404 on missing plugin', async (t) => {
  const req = httpMocks.createRequest({ params: { plugin: '---missing-plugin' } });
  req.db = new DB();
  const res = httpMocks.createResponse();

  await issuesRoute(req, res);

  t.is(404, res.statusCode);
});
