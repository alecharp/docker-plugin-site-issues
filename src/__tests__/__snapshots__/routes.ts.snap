// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes GET /issues/:plugin/open should return 200 1`] = `
Array [
  Array [
    Object {
      "issues": Array [
        Object {
          "assignee": "",
          "created": "2021-03-29T05:04:57Z",
          "issueType": "dependencies",
          "key": "#1588",
          "priority": "",
          "reporter": "dependabot[bot]",
          "resolution": "open",
          "status": "open",
          "summary": "Bump artifactory from 3.4.0 to 3.10.6",
          "updated": "2021-03-29T05:05:00Z",
          "url": "https://github.com/jenkinsci/configuration-as-code-plugin/pull/1588",
        },
        Object {
          "assignee": "",
          "created": "2021-03-22T05:02:03Z",
          "issueType": "dependencies",
          "key": "#1585",
          "priority": "",
          "reporter": "dependabot[bot]",
          "resolution": "open",
          "status": "open",
          "summary": "Bump aws-java-sdk from 1.11.457 to 1.11.976",
          "updated": "2021-03-22T05:02:07Z",
          "url": "https://github.com/jenkinsci/configuration-as-code-plugin/pull/1585",
        },
        Object {
          "created": "2021-01-19T23:32:27.000+0000",
          "issueType": "Bug",
          "key": "JENKINS-64664",
          "priority": "Minor",
          "reporter": "Kevin Carrasco",
          "summary": "Kubernetes plugin jCasC compatibility issue",
          "updated": "2021-02-05T09:42:22.000+0000",
          "url": "https://issues.jenkins.io/browse/JENKINS-64664",
        },
      ],
    },
  ],
]
`;

exports[`routes GET /issues/:plugin/open should return 200 2`] = `Array []`;

exports[`routes GET /issues/:plugin/open should return 200 3`] = `Array []`;

exports[`routes GET /issues/:plugin/open should return 404 on missing plugin 1`] = `Array []`;

exports[`routes GET /issues/:plugin/open should return 404 on missing plugin 2`] = `
Array [
  Array [
    "No such plugin",
  ],
]
`;

exports[`routes GET /issues/:plugin/open should return 404 on missing plugin 3`] = `Array []`;
